upgrade/upgrades: TestWaitForDelRangeInGCJob failed

upgrade/upgrades.TestWaitForDelRangeInGCJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9475758?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9475758?buildTab=artifacts#/) on master @ [7d219406b9c0fc5079403840d55ba40f721512f9](https://github.com/cockroachdb/cockroach/commits/7d219406b9c0fc5079403840d55ba40f721512f9):


```
=== RUN   TestWaitForDelRangeInGCJob
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob776415057
    test_log_scope.go:79: use -show-logs to present logs inline
    wait_for_del_range_in_gc_job_test.go:117: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/wait_for_del_range_in_gc_job_test.go:117
        	Error:      	"[{/Table/106/1/1/0 raw_bytes:"\231\354\326\303\n" timestamp:<wall_time:1680806404058785775 > }]" should have 0 item(s), but has 1
        	Test:       	TestWaitForDelRangeInGCJob
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob776415057
--- FAIL: TestWaitForDelRangeInGCJob (12.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWaitForDelRangeInGCJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26653

upgrade/upgrades.TestWaitForDelRangeInGCJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9477538?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9477538?buildTab=artifacts#/) on master @ [af563f8897aac5c8751566345e83c2c008a887dc](https://github.com/cockroachdb/cockroach/commits/af563f8897aac5c8751566345e83c2c008a887dc):


```
=== RUN   TestWaitForDelRangeInGCJob
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob517025222
    test_log_scope.go:79: use -show-logs to present logs inline
    wait_for_del_range_in_gc_job_test.go:117: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/wait_for_del_range_in_gc_job_test.go:117
        	Error:      	"[{/Table/106/1/1/0 raw_bytes:"\231\354\326\303\n" timestamp:<wall_time:1680811893269579391 > }]" should have 0 item(s), but has 1
        	Test:       	TestWaitForDelRangeInGCJob
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob517025222
--- FAIL: TestWaitForDelRangeInGCJob (13.80s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWaitForDelRangeInGCJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestWaitForDelRangeInGCJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9479924?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9479924?buildTab=artifacts#/) on master @ [63f32874ed846c4d2913dc9efab881c2aa979308](https://github.com/cockroachdb/cockroach/commits/63f32874ed846c4d2913dc9efab881c2aa979308):


```
=== RUN   TestWaitForDelRangeInGCJob
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob3257028233
    test_log_scope.go:79: use -show-logs to present logs inline
    wait_for_del_range_in_gc_job_test.go:117: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/wait_for_del_range_in_gc_job_test.go:117
        	Error:      	"[{/Table/106/1/1/0 raw_bytes:"\231\354\326\303\n" timestamp:<wall_time:1680822046202430760 > }]" should have 0 item(s), but has 1
        	Test:       	TestWaitForDelRangeInGCJob
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob3257028233
--- FAIL: TestWaitForDelRangeInGCJob (6.44s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWaitForDelRangeInGCJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestWaitForDelRangeInGCJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9479924?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9479924?buildTab=artifacts#/) on master @ [63f32874ed846c4d2913dc9efab881c2aa979308](https://github.com/cockroachdb/cockroach/commits/63f32874ed846c4d2913dc9efab881c2aa979308):


```
=== RUN   TestWaitForDelRangeInGCJob
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob939030977
    test_log_scope.go:79: use -show-logs to present logs inline
    wait_for_del_range_in_gc_job_test.go:117: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/wait_for_del_range_in_gc_job_test.go:117
        	Error:      	"[{/Table/106/1/1/0 raw_bytes:"\231\354\326\303\n" timestamp:<wall_time:1680822116625565528 > }]" should have 0 item(s), but has 1
        	Test:       	TestWaitForDelRangeInGCJob
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestWaitForDelRangeInGCJob939030977
--- FAIL: TestWaitForDelRangeInGCJob (7.25s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWaitForDelRangeInGCJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/wait_for_del_range_in_gc_job_test.go
+++ b/pkg/upgrade/upgrades/wait_for_del_range_in_gc_job_test.go
@@ -21,10 +21,12 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
 
@@ -110,18 +112,31 @@ SELECT count(*)
 	tdb.Exec(t, "SET CLUSTER SETTING version = crdb_internal.node_executable_version()")
 
 	codec := testServer.ExecutorConfig().(sql.ExecutorConfig).Codec
-	{
-		dbFooStart := codec.TablePrefix(dbFooID)
-		res, err := kvDB.Scan(ctx, dbFooStart, dbFooStart.PrefixEnd(), 1)
-		require.NoError(t, err)
-		require.Len(t, res, 0)
-	}
-	{
-		idxStart := codec.IndexPrefix(fooID, idxID)
-		res, err := kvDB.Scan(ctx, idxStart, idxStart.PrefixEnd(), 1)
-		require.NoError(t, err)
-		require.Len(t, res, 0)
-	}
+	testutils.SucceedsSoon(t,
+		func() error {
+			{
+				dbFooStart := codec.TablePrefix(dbFooID)
+				res, err := kvDB.Scan(ctx, dbFooStart, dbFooStart.PrefixEnd(), 1)
+				if err != nil {
+					return err
+				}
+				if len(res) != 0 {
+					return errors.AssertionFailedf("unexpected number of table keys (got %d)", len(res))
+				}
+			}
+			{
+				idxStart := codec.IndexPrefix(fooID, idxID)
+				res, err := kvDB.Scan(ctx, idxStart, idxStart.PrefixEnd(), 1)
+				require.NoError(t, err)
+				if err != nil {
+					return err
+				}
+				if len(res) != 0 {
+					return errors.AssertionFailedf("unexpected number of index keys (got %d)", len(res))
+				}
+			}
+			return nil
+		})
 
 	// Make sure that there is still MVCC history.
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM foo@idx AS OF SYSTEM TIME "+beforeDrop,

